src/core/fem/src/condition/4C_fem_condition_multipointcoupling.cpp

```diff
@@ -0,0 +1,243 @@
+  }
+
+  // redistribute such that all master and corresponding slave nodes are owned by the same processor
+  redistribute(rank_to_hold_condition);
+}
+
+void Core::Conditions::MultipointcouplingConditions::redistribute(
+    std::shared_ptr<std::vector<int>> rank_to_hold_condition)
+{
+  // make sure we have a filled discretisation at this place
+  // dofs are not required yet, they are assigned after redistribution
+  // accessing the noderowmap requires a 'completed' discretization
+  if (!discret_->filled())
+  {
+    discret_->fill_complete(false, false, false);
+  }
+
+  // a list of all nodes on this proc
+  std::vector<int> nodesonthisproc(discret_->node_row_map()->NumMyElements());
+
+  // get all owned node gids of this proc
+  discret_->node_row_map()->MyGlobalElements(nodesonthisproc.data());
+  std::set<int> row_node_set(nodesonthisproc.begin(), nodesonthisproc.end());
+
+  // move all conditioned nodes to corresponding rank
+  int myerase = 0;
+  int numerase = 0;
+  int myadd = 0;
+  int numadd = 0;
+
+  for (size_t i = 0; i < all_pointcoupling_conditions_.size(); ++i)
+  {
+    auto& current_condition = all_pointcoupling_conditions_[i];
+    const std::vector<int>* conditioned_nodes = current_condition->get_nodes();
+
+    // add conditioned nodes to target processor and remove from others
+    if ((*rank_to_hold_condition)[i] == myrank_)
+    {
+      const int size_before = static_cast<int>(row_node_set.size());
+      row_node_set.insert(conditioned_nodes->begin(), conditioned_nodes->end());
+      myadd += static_cast<int>(row_node_set.size()) - size_before;
+    }
+    else
+    {
+      const int size_before = static_cast<int>(row_node_set.size());
+      for (int idtodel : (*conditioned_nodes))
+      {
+        row_node_set.erase(idtodel);
+      }
+      myerase -= static_cast<int>(row_node_set.size()) - size_before;
+    }
+  }
+
+  Core::Communication::sum_all(&myerase, &numerase, 1, discret_->get_comm());
+  Core::Communication::sum_all(&myadd, &numadd, 1, discret_->get_comm());
+  if (myrank_ == 0 && verbose_)
+  {
+    std::cout << "Erased " << numerase << " nodes in total from row node list.\n";
+    std::cout << "Added " << numadd << " nodes in total from row node list.\n";
+  }
+
+  {
+    // safety check
+    int myn = static_cast<int>(row_node_set.size());
+    int gn = 0;
+
+    Core::Communication::sum_all(&myn, &gn, 1, discret_->get_comm());
+    if (gn != discret_->num_global_nodes())
+    {
+      FOUR_C_THROW(
+          "Unmatching numbers of nodes before and after call Redistribution. Nodemap "
+          "constructor will crash.\n");
+    }
+  }
+
+  nodesonthisproc.clear();
+  std::vector<int> newnodesonthisproc(row_node_set.begin(), row_node_set.end());
+
+  //--------------------------------------------------
+  // build noderowmap for new distribution of nodes
+  // the constructed rowmap
+  std::shared_ptr<Core::LinAlg::Map> newrownodemap =
```

Comment:
This shared_ptr seems unnecessary. Same for `newcolnodemap`. Also, what do you think about naming this `new_row_node_map` etc?